utils/src/server/rpc_module.rs

```diff
@@ -88,61 +89,84 @@ impl RpcModule {
 		Ok(())
 	}
 
-	/// Register a new RPC subscription, with subscribe and unsubscribe methods.
-	pub fn register_subscription(
+	/// Register a new RPC subscription that invokes callback on every subscription request.
+	/// The callback itself takes two parameters:
+	///     - RpcParams: JSONRPC parameters in the subscription request.
+	///     - SubscriptionSink: A sink to send messages to the subscriber.
+	///
+	/// # Examples
+	///
+	/// ```no_run
+	///
+	/// use jsonrpsee_utils::server::rpc_module::RpcModule;
+	///
+	/// let mut rpc_module = RpcModule::new();
+	/// rpc_module.register_subscription("sub", "unsub", |params, sink| {
+	///     let x: usize = params.one()?;
+	///     std::thread::spawn(move || {
+	///         sink.send(&x)
+	///     });
+	///     Ok(())
+	/// });
+	/// ```
+	pub fn register_subscription<F>(
 		&mut self,
 		subscribe_method_name: &'static str,
 		unsubscribe_method_name: &'static str,
-	) -> Result<SubscriptionSink, Error> {
+		callback: F,
+	) -> Result<(), Error>
+	where
+		F: Fn(RpcParams, SubscriptionSink) -> Result<(), Error> + Send + Sync + 'static,
+	{
 		if subscribe_method_name == unsubscribe_method_name {
 			return Err(Error::SubscriptionNameConflict(subscribe_method_name.into()));
 		}
 
 		self.verify_method_name(subscribe_method_name)?;
 		self.verify_method_name(unsubscribe_method_name)?;
 
-		let subscribers = Arc::new(Mutex::new(FxHashMap::default()));
-
 		{
-			let subscribers = subscribers.clone();
+			let subscribers = self.subscribers.clone();
 			self.methods.insert(
 				subscribe_method_name,
-				Box::new(move |id, _, tx, conn| {
+				Box::new(move |id, params, method_sink, conn| {
+					let (online_tx, online_rx) = oneshot::channel::<()>();
 					let sub_id = {
 						const JS_NUM_MASK: SubscriptionId = !0 >> 11;
-
 						let sub_id = rand::random::<SubscriptionId>() & JS_NUM_MASK;
 
-						subscribers.lock().insert((conn, sub_id), tx.clone());
+						subscribers.lock().insert((conn, sub_id), (method_sink.clone(), online_rx));
```

Comment:
do we want to check that a `already existing` subscription is not overwritten here?

Should be very very very unlikely but raising the question.